ykval-checksum-clients.1
.\" Copyright (c) 2011-2015 Yubico AB
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions are
.\" met:
.\"
.\"     * Redistributions of source code must retain the above copyright
.\"       notice, this list of conditions and the following disclaimer.
.\"
.\"     * Redistributions in binary form must reproduce the above
.\"       copyright notice, this list of conditions and the following
.\"       disclaimer in the documentation and/or other materials provided
.\"       with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
.\" "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
.\" LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
.\" A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
.\" OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
.\" SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
.\" LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
.\" OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.\" The following commands are required for all man pages.
.de URL
\\$2 \(laURL: \\$1 \(ra\\$3
..
.if \n[.g] .mso www.tmac
.TH ykval-checksum-clients "1" "January 2013" "yubico-val"
.SH NAME
ykval-checksum-clients - Calculate a checksum of all Client Info data.
.SH SYNOPSIS
.B ykval-checksum-clients
.SH DESCRIPTION
Calculates a checksum using the id, active, and secret fields of all Client 
Info data in the yubikey-val server database. This checksum can be used to 
easily compare the state of clients in two yubikey-val servers in the same 
sync pool.

Database configuration is read from /etc/yubico/val/ykval-config.php
.SH BUGS
Report ykval-checksum-clients bugs in
.URL "https://github.com/Yubico/yubikey-val/issues" "the issue tracker"
.SH "SEE ALSO"
.BR ykval-checksum-deactivated (1)

The
.URL "https://github.com/Yubico/yubikey-val" "yubikey-val home page"
.PP
YubiKeys can be obtained from
.URL "https://www.yubico.com/" "Yubico" "."



NEWS
* Version 2.30 (unreleased)

* Version 2.29 (released 2015-05-27)

  * Allow curl options to be set from config file.

* Version 2.28 (released 2015-02-11)

  * Refactor munin quelenegth plugin to show what is queued.

  * Add ykval-nagios-queuelength.

  * Use constant time string comparisson for validating HMAC signature.

* Version 2.27 (released 2014-09-25)

  * Further logging updates.

  * ykval-munin-responses: Make log file configurable.

  * ykval-munin-ksmresponses: New munin probe.

* Version 2.26 (released 2014-09-24)

  * Logging updates.

  * Optimization fix in the checksum scripts.

  * Documentation fixes.

* Version 2.25 (released 2014-08-18)
  
  * Now works with 'allow_url_fopen' == false.

  * Always verifies SSL peer when syncing between servers via HTTPS.

* Version 2.24 (released 2013-09-18)
  
  * Removed space after comma in the output of ykval-gen-clients.

  * Include README in tarball.

* Version 2.23 (released 2013-04-17)

  * Removed initial empty line from output for all commands.

  * Use LF as EOL consistently.

  * Updated release procedure.

* Version 2.22 (released 2013-03-12)

  * Added the ability to send yk=all to ykval-resync.php to queue sync
    of all known active YubiKeys.

  * Added ykval-synchronize to easily call ykval-resync.php on a remote
    server.

  * Added ykval-gen-clients to generate API clients.

  * Log query for POST requests too.

* Version 2.21 (released 2013-02-05)

  * Fixed a problem that caused ykval-queue to terminate if the database
    was not available initially.

* Version 2.20 (released 2013-01-31)

  * Add ChangeLog generation using git2cl.

  * Changed location of files to /usr/share/yubikey-val, etc.

  * Changed location of configuration files to /etc/yubico/val/.

  * Made import/export scripts use comma separation, instead of tabs.

  * Added a working ykval-config.php that looks for a ksm on localhost.

  * Removed System_Daemon dependency and made ykval-queue a simple
    backgroundable process that can be daemonized using for instance
    and init.d script.

  * Added man pages for executables.

* Version 2.19 (released 2012-07-05)

  * Refactor database code, allowing for other underlying implementations
    than PDO. Add a PDO and an Oracle (through php_oci) implementation.
    Based on patch from Remi Mollon <remi.mollon@cern.ch>

  * Fix for ykval-export running on postgres.

  * Add resync.php to request new sync of public id.

  * Add munin plugin for statistics.

* Version 2.18 (released 2012-06-15)

  * Logging misstakes that broke 2.17 fixed.

* Version 2.17 (released 2012-06-15)

  * Logging improvements.
    use ykval-verify/ykval-sync correctly for whole flow
    clarify/degrade various logging messages

  * Fix mysql error introduced in 2.14, also logs
    database updated/not updated correctly.

  * Accept sync for disabled keys, but still answer BAD_OTP.

  * Remove from sync queue on BAD_OTP answer.

  * Add munin plugin for response types.

* Version 2.16 (released 2012-06-13)

  * Improved logging.

  * Improved performance of large sync queues.

* Version 2.15 (released 2012-05-24)

  * Add export/import scripts for clients table.

  * Insert default values in $sl and $timeout if they are empty.
    And they will be empty if the client didn't request them.

* Version 2.14 (released 2012-05-22)

  * Add support for reconnecting to database after errors.

  * Fixes for PHP warnings.

  * Detect timeouts and errors in munin checks.

* Version 2.13 (released 2012-05-16)

  * Fix signature checking broken in 2.12 and for dvorak OTPs.

  * Fixes for ykval-checksum-clients.php

* Version 2.12 (released 2012-05-09)

  * Fix using 'fast' or 'secure' as sync level.

  * Fix database setup script to make nonce max 40 characters.

* Version 2.11 (released 2011-11-16)

  * Silence PHP warnings.  Patch from Hiroki Nose.

  * Include munin scripts in tarball.  From Fredrik Thulin.

  * Support for DESTDIR in 'make install'.  From Fredrik Thulin.

  * Reorder include's to allow for dbi-settings through
    ykval-config.php.  From Fredrik Thulin.

  * Install non-bin PHP files with --mode 644 to avoid executable bit.
    From Fredrik Thulin.

  * Fix two remaining non-portable uses of rowCount.

* Version 2.10 (released 2011-08-18)

  * Don't echo (unsanitized) OTP/NONCE values back to client when
    sending error codes.  Reported by Paul van Empelen.

    Resolving this problem protects (arguably buggy) clients against
    an attack.  Prior versions of the Yubico C and PHP clients do not
    appear to exhibit this bug.  We provide an analysis of the issue
    below so that you can review client implementations for the
    problem.  Note that you do not have to fix clients if you are
    using this server version (or later), although we recommend it
    anyway.

    If the client sends a OTP value that ends with '%0astatus=OK' the
    server output will contain a line 'status=ok' before the real
    status code status=MISSING_PARAMETER.  Note lower-casing of the
    injected status code, so that it doesn't match a correct
    'status=OK' response.  Note also that the OTP value would fail
    normal input validation checks in the client.

    If the client sends a NONCE value that ends with '%0astatus=OK'
    the output will contain a line consisting of 'status=OK' before
    the correct status=MISSING_PARAMETER.  However, the NONCE value is
    generated by client code internally and does not come from any
    untrusted source, thus the impact here is limited -- if an
    attacker is able to trick a client into sending a crafted NONCE
    value the attacker is normally able to modify the client code
    somehow, and can thus trick the client in other ways as well.
    Similar issues apply to the ID field, which is normally also under
    control of the trusted client code and not something an attacker
    could influence.

    Thus, this server-side fix solve a client-side issue that we

ykval-checksum-clients.php
#!/usr/bin/php
<?php

$verbose = 0;
if (isset($argv[1])) {
  if ($argv[1] == "-h" || $argv[1] == "--help") {
    print "Usage: " . $argv[0] . " [-h|--help] [-v]\n";
    exit(1);
  }

  if ($argv[1] && $argv[1] != "-v") {
    print $argv[0] . ": invalid option -- '" . $argv[0] . "'\n";
    print "Try `" . $argv[0] . " --help' for more information.\n";
    exit(1);
  }

  $verbose = $argv[1] == "-v";
}

set_include_path(get_include_path() . PATH_SEPARATOR .
		 "/usr/share/ykval:/etc/ykval");

require_once 'ykval-config.php';
require_once 'ykval-db.php';

$logname="ykval-checksum-clients";
$myLog = new Log($logname);

$db = Db::GetDatabaseHandle($baseParams, $logname);

if (!$db->connect()) {
  $myLog->log(LOG_WARNING, "Could not connect to database");
  exit(1);
}

$everything = "";
$result=$db->customQuery("SELECT id, active, secret ".
			 "FROM clients ".
			 "ORDER BY id");
while($row = $db->fetchArray($result)) {
  $active = $row['active'];
  if ($active == "") {
    # For some reason PostgreSQL returns empty strings for false values?!
    $active = "0";
  }
  $everything = $everything .
    $row['id'] . "\t" . $active . "\t" .
    $row['secret'] . "\n";
}

$db->closeCursor($result);
$hash = sha1 ($everything);

if ($verbose) {
  print $everything;
}
print substr ($hash, 0, 10) . "\n";

$result=null;
$db=null;

?>
